Monitor wifi connectivity

jjjasont opened the pull request:
Add a new process for Wifi connectivity check.

If SSID and WIFI_MAX_FAILS are defined, Wi-Fi monitoring is enabled.

Teslausb will check wifi connectivity on a 60-seconds internal.
Teslausb will reboot if SSID is in range, but not connected after WIFI_MAX_FAILS

jjjasont committed:
Update archiveloop

new process to monitor wifi connectivity

jjjasont committed:
Update teslausb_setup_variables.conf.sample

wifi connectivity monitor teslausb_setup_variables.conf parameters

jjjasont commented:
Background for this pull request:

I setup Home wifi SSID and enabled AP access point in TeslaUSB.

I noticed that TeslaUSB does not automatically connects to home wifi when I return home. But it will successfully connect to home wifi after a reboot.

So I setup a monitor script to reboot TeslaUSB if after X consecutive failed checks

jjjasont committed:
Update archiveloop

Do a ping test connectivity.
Default to ping 8.8.8.8 if ping_host not defined in config

jjjasont committed:
Update teslausb_setup_variables.conf.sample

Update with ping host config

jjjasont committed:
Update teslausb_setup_variables.conf.sample

typo

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -16,6 +16,13 @@
 export SSID='your_ssid'
 export WIFIPASS='your_pass'
 
+# Wi-Fi connectivity check settings (optional)
+# Wi-Fi monitoring requires SSID and MAX_FAILS; PING_HOST is optional
+# TeslaUSB will check wifi connectivity on a 60-seconds internal.
```
The new comment is inconsistent: "Wi-Fi" vs. "wifi".
Existing lines in this file use "wifi", so please just use that (capitalized to "Wifi" at the start of a sentence).

jjjasont replied:
This is updated.

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -16,6 +16,13 @@
+# Wi-Fi connectivity check settings (optional)
+# Wi-Fi monitoring requires SSID and MAX_FAILS; PING_HOST is optional
+# TeslaUSB will check wifi connectivity on a 60-seconds internal.
+# TeslaUSB will reboot if SSID is in range, but ping failed after WIFI_MAX_FAILS
+#export MAX_FAILS  = 3
```
You renamed the variable to MAX_FAILS in a 2nd commit, but did not update the comment to match.
I prefer WIFI_MAX_FAILS over MAX_FAILS though.

jjjasont replied:
Agreed.
Variable revert back to WIFI_MAX_FAILS

marcone commented on pi-gen-sources/00-teslausb-tweaks/files/teslausb_setup_variables.conf.sample:
```diff
@@ -16,6 +16,13 @@
+# TeslaUSB will check wifi connectivity on a 60-seconds internal.
+# TeslaUSB will reboot if SSID is in range, but ping failed after WIFI_MAX_FAILS
+#export MAX_FAILS  = 3
+#export PING_HOST="192.168.1.1"    # Host to ping for connectivity; defaults to 8.8.8.8 if unset
```
I feel like this should either ping the archive server, or the default gateway.
I think I prefer the latter, because then it would do the right thing (which is nothing) both if the inernet is down or if the archive server is down.

jjjasont replied:
Do you think this is just a change in the comment?
The current version, the user can specify a host that response to ping. It can be router/gateway/archive server or external ip address.

marcone replied:
> Do you think this is just a change in the comment?

I was thinking that the code that actually does the pinging should default to the gateway if PING_HOST is unspecified, i.e. something like `: ${PING_HOST:=$(route | grep '^default' | awk '{print $2}')}` and then the user can just leave it unspecified in the config file. That would then also work if the user specified multiple networks to connect to (e.g. home, work, hotspot, etc.)

marcone commented on run/archiveloop:
```diff
@@ -824,6 +824,45 @@ function wifichecker {
+    # Check only if both SSID and MAX_FAILS are defined and non-empty
+    if [ -n "$SSID" ] && [ -n "$MAX_FAILS" ]; then
+      # Force a fresh Wi-Fi scan
+      nmcli dev wifi rescan >/dev/null 2>&1 || log "Wi-Fi rescan failed"
```
The system TeslaUSB is running on may not be using Network Manager, so this code should either be changed to work with both Network Manager and wpa_supplicant, or it should only be run if Network Manager is used.

jjjasont replied:
Updated to handle both Network Manager & wpa_supplicant.

marcone commented on run/archiveloop:
```diff
@@ -824,6 +824,45 @@ function wifichecker {
+      : "${PING_HOST:=8.8.8.8}"
+
+      # Check if $SSID is in range using terse output
+      if nmcli -t -f SSID dev wifi list | grep -q "^$SSID$"; then
```
TeslaUSB code style is:
```
if condition
then
 ...
```

jjjasont replied:
Thanks. Done.

jjjasont committed:
Update archiveloop

Rename MAX_FAILS back to WIFI_MAX_FAILS.
Support both Networking Manager & WPA Supplicant.
Update the monitoring logic so that monitoring will only happen if SSID, WIFI_MAX_FAILS and PING_HOST are defined.

jjjasont committed:
Update teslausb_setup_variables.conf.sample

Rename MAX_FAILS back to WIFI_MAX_FAILS.
Update descriptions on how wifi monitoring works.

jjjasont committed:
Update archiveloop

1. Network_ID match needs to be an exact match. Not partial match.
2. PING_HOST to be optional. If PING_HOST not specified, ping the default gateway.

jjjasont committed:
Update teslausb_setup_variables.conf.sample

1. Update the comment on PING_HOST

jjjasont committed:
Update archiveloop

Add log message

jjjasont committed:
Update archiveloop

This is a working version.
To finetune the reboot logic in the next version.
To clean up the logs and comments in the next version.

jjjasont committed:
Merge branch 'marcone:main-dev' into monitor_wifi_connectivity

jjjasont committed:
Update archiveloop

Complete rewrite of the function.
Simpler logic flow
Also included a wifi disconnect monitoring after connected to wifi.

jjjasont committed:
Update teslausb_setup_variables.conf.sample

Update config file
